### PR TITLE
Signed session IDs and consistent session/url/app_path params

### LINKED ISSUE #3171: `pull_session` should let you provide app_path, host, port
It's a pain to construct / figure out the websocket url for pull_session:
https://github.com/bokeh/bokeh/blob/d06b4dd27402921bfb3d73d08e4b785c774ee280/bokeh/client/session.py#L43

It would be more logical I think to provide app_path, host, and port. Or provide a utility function that assembles a websocket url from server url and app_path... maybe `websocket_url_for_server_url` could take a base server url, and an app_path ? https://github.com/bokeh/bokeh/blob/d06b4dd27402921bfb3d73d08e4b785c774ee280/bokeh/resources.py#L30

### LINKED ISSUE #3324: autoload_server should not use default session id
session_id='default' is an awful default because apps will put all users in the same session by default,
which will make production deployments explode.

### PR DESCRIPTION
**Pull request by havocp**
This is intended to fix #3324 and fix #3171 and also fix output_server to have an app_path which is in the wiki (https://github.com/bokeh/bokeh/wiki/Issue-list:-new-Bokeh-server ) but not an open issue.

Signed sessions are a partial step toward https://github.com/bokeh/bokeh/issues/3287 but there's a missing piece (doc_handler.py will still autogenerate an ID, there has to be a way to disable that).

### CODE REVIEW

**Comment by bryevdv** [comment]
stopped the first build to free up workers

**Comment by bryevdv** [comment]
@havocp I merged master, I had to change a few `ws_url` to `url` which seemed intentional, but please double check. What else is left to do for this PR?

**Comment by havocp** [comment]
this could be reviewed as is but full utility really needs a server mode that refuses to gen session ids on server side. I was also thinking about whether the env vars should be fallbacks overridable via command line option. Another nice addition would be a bokeh sub command to gen a secret key for you. Anyway those things can go here or in a future PR.

**Comment by bryevdv** [comment]
Let's make issues and put them in other PRs I think.

**Comment by havocp** [comment]
cool. this PR has high potential to break something or get the crypto wrong I think so careful review appreciated.

File: examples/embed/widget.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ def update_pyramid(self):
 pop = Population()
 pop.render()
 
-tag = embed.autoload_server(pop.layout)
+tag = embed.autoload_server(session_id=pop.session.id, pop.layout)
```
Comment (bryevdv) [inline examples/embed/widget.py]: I understand why this is the way it is but I wonder if there is any additional convenience layer we could add that would free the user from having to make this coordination. `session.autoload_server(layout)`, `embed.autoload_from_cursession(layout)`?

Comment (havocp) [inline examples/embed/widget.py]: a bokeh.io.autoload could grab the session id, but for these examples might need an output_server(keep_session_open) or something ...

File: bokeh/util/session_id.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
+''' Provide a request handler that returns a page displaying a document.
+
+'''
```
Comment (bryevdv) [inline bokeh/util/session_id.py]: Docstring needs update

File: examples/embed/widget.py (hunk 1 above)
Comment (havocp) [inline examples/embed/widget.py]: though, I just realized in two of the most common cases this doesn't arise.
 * with output_notebook, we autoload_server for you: https://github.com/bokeh/bokeh/pull/3325/files#diff-ab5e8987987503b1248b3f5c707e056dR270
 * if you were writing an app that embeds an autoload, you only have to pull_session if you are going to modify each session; the more common case is to just call `autoload_server` and let it generate a session_id for you, with no session object seen

Comment (bryevdv) [inline examples/embed/widget.py]: OK that seems fine. It would be valuable to codify these situations/use-cases/guideance somewhere.

Comment (havocp) [inline examples/embed/widget.py]: I could add a little more info to some of the doc strings, will give that a shot.

**Comment by bryevdv** [comment]
So this LGTM to the extent of my ability to evaluate it. I tested it and everything works to the level expected (some fixes in other PRs, etc)

**Comment by havocp** [comment]
Filed
* don't autogen session ids in server https://github.com/bokeh/bokeh/issues/3343
* add `bokeh secret` https://github.com/bokeh/bokeh/issues/3344

**Comment by havocp** [comment]
I think the test failure is bogus. this one should be ready to review/merge.

**Comment by bryevdv** [comment]
I'll merge when travis is green

**Pull request merged**